### PR TITLE
Pparse: check ast invariants after preprocessing

### PR DESCRIPTION
**Pull request by Octachron**
This small PR proposes to check AST invariants after the preprocessing phase of the toplevel.

For instance, running a faulty ppx that generates an empty tuple with
```OCaml
let t = [%tuple];;
```
raises with this PR an error in the toplevel itself:
> Error: broken invariant in parsetree: Tuples must have at least 2 components.

rather than an internal fatal error:
> Fatal error: exception File "typing/typecore.ml", line 2999, characters 6-12: Assertion failed

### CODE REVIEW

**Review by gasche — APPROVED** [review]
We have discussed this before, and I agree it is a good idea. The implementation looks good.

**Comment by Octachron** [comment]
I added a change entry (as a bug fix) since the change is potentially observable by users (or, more probably, for ppx developpers). I will merge once CI pass.

**Comment by gasche** [comment]
I would placed the change entry in "Tools" (toplevel) or "Internal/compiler-libs changes" (where -ppx authors may also go look), but whatever.

**Comment by alainfrisch** [comment]
Shouldn't this logic go to Pparse, to increase the likelihood that external tool also using the ppx infrastructure from compiler-libs would apply the same checks (assuming it is a good idea that they do!)?

**Comment by gasche** [comment]
There already is logic in Pparse, but only in `parse_file` I think that the toplevel does not reuse. You're suggesting to push the checks in `apply_rewriters`, I suppose. I'll let @Octachron decides what he thinks is best.

**Comment by alainfrisch** [comment]
Yes, exactly.  On a slightly related note, the treatment of ImplementationHooks.apply_hooks seems suspicious to me: sometimes those hooks are applied before type-checking, sometimes after.

**Comment by Octachron** [comment]
From the point of view of external  tools, I am not sure if adding an `Ast_invariant` check at the end of `apply_rewriters` *always* make sense, since the final consumer might not be the compiler?

**Comment by alainfrisch** [comment]
> since the final consumer might not be the compiler?

It's a rather abstract discussion, but I view invariants as things that one would like to express in the Parsetree definition, but one cannot do it due to technical limitations (or it would make the code too heavy/irregular).  Not particularly as expectations specific to the compiler.  This is also reflected in the fact that attribute payloads are checked for invariants, even if the compiler doesn't do anything with them.

One expect in particular some kind of roundtripping property between the parser and printers (in concrete syntax), limited to Parsetree that respects the invariants.  And this should hold also for "ppx" called by external tools, for debugging purposes.

**Comment by Octachron** [comment]
Fair enough. I am moving the invariant check to `apply_rewriters`.

**Comment by Octachron** [comment]
I have moved the invariants tests to `Pparse.apply_rewriters` and the change entry to the `compiler-libs` section.

File: driver/pparse.ml
Comment (gasche) [inline driver/pparse.ml]: I find the logic for this line here fairly obscure. Could you just move the `invariant_fun` call to within the `is_ast` branch, so that it would test the invariants before the ppx extensions are run on the file? (This is assuming that checking the invariants is cheap, so that doing it twice is not a big deal.)

Comment (gasche) [reply driver/pparse.ml]: (Or if we don't want to check both before and after rewriting, you can still condition on `all_ppx = []` within the branch.)

Comment (Octachron) [reply driver/pparse.ml]: Maybe adding a comment on the logic and moving it to the `is_ast` branch is sufficient?
```OCaml
let ast = (input_value ic : a) in
  if !Cflags.all_ppx = [] then invariant_fun ast;
  (* if all_ppx <> [], invariant_fun will be called by apply_rewriters *)
  ast
```

Comment (gasche) [reply driver/pparse.ml]: Yes, it seems reasonable to me.

Comment (Octachron) [reply driver/pparse.ml]: Done.

**Comment by gasche** [comment]
I'm not sure whether the testsuite still passes with the changes to location-printing that may have hit trunk in between. @Octachron, could you do a rebase before I merge?

**Comment by Octachron** [comment]
Sure

**Comment by gasche** [comment]
(As far as I can tell the HEAD has not changed and the CI is not rerun yet. I'm not sure if your message above means "I'll do the rebase later" or "I did the rebase but forgot to push".)

**Comment by Octachron** [comment]
It was the first option, the PR is now rebased.

**Pull request merged**
